### ISSUE TITLE
Channel numbers may be re-used before they are ready

### ISSUE THREAD

**Issue by nkvoll** [pika/pika#84]
Example:

```
import pika
from pika import exceptions
from pika.adapters import twisted_connection
from twisted.internet import defer, reactor, protocol


def wait(n=0):
    """ A simple asynchronous sleep. """
    d = defer.Deferred()
    reactor.callLater(n, d.callback, None)
    return d

@defer.inlineCallbacks
def run(connection):
    # open a channel and declare a queue:
    channel = yield connection.channel()
    print 'got channel:', channel, channel.channel_number
    queue = yield channel.queue_declare(queue='foobar', auto_delete=True, exclusive=True)

    for i in range(10):
        current_channel = yield connection.channel()

        print 'current channel:', current_channel.channel_number, current_channel
        try:
            # create a queue with incompatible arguments:
            bad_queue = yield current_channel.queue_declare(queue='foobar', auto_delete=False, exclusive=True)
        except exceptions.ChannelClosed as cc:
            print 'Channel closed:', cc 
            # waiting for 0 seconds means wait 1 reactor iteration
            yield wait(0)

    reactor.stop()

parameters = pika.ConnectionParameters()    
cc = protocol.ClientCreator(reactor, twisted_connection.TwistedProtocolConnection, parameters)

d = cc.connectTCP('localhost', 5672)
d.addCallback(lambda protocol: protocol.ready)
d.addCallback(run)

reactor.run()
```

The output from the above script:

```
$ python ~/tmp/pika_test.py
got channel: <pika.adapters.twisted_connection.TwistedChannel object at 0x10143d190> 1
current channel: 2 <pika.adapters.twisted_connection.TwistedChannel object at 0x10143d490>
Channel closed: (406, "PRECONDITION_FAILED - parameters for queue 'foobar' in vhost '/' not equivalent")
/Users/username/.virtualenvs/Piped/lib/python2.7/site-packages/pika/callback.py:69: UserWarning: CallbackManager.add: Duplicate callback found for "2:Channel.CloseOk"
  (self.__class__.__name__, prefix, key))
<< the program hangs here, waiting on line 21 (yield connection.channel()) >>
```

The reason seems to be that the channel number is reused before it actually freed, my guess is because the 406 error from the server isn't handled properly.

I've noticed that this may sometimes manifest in the following error (being sent to connection.close):
    503 COMMAND_INVALID - second 'channel.open' seen
